liferay-frontend/liferay-portal
# LPS-169939 - Research ways to use Client extensions as dependencies of other extensions

This is the PoC for testing the [proposed feature](https://issues.liferay.com/browse/LPS-169939). 

Steps to tests:

> ⚠️ Note that you may skip steps 3-8 and deploy the pre-built [greeter.zip](https://github.com/liferay-frontend/liferay-portal/files/10435289/greeter.zip) and [provider.zip](https://github.com/liferay-frontend/liferay-portal/files/10435290/provider.zip) files if you just want to see the solution working in your local portal instance (and don't want to mess with the blade workspace).
>
> Nevertheless, it's recommended to have a look at `greeter` and `provider` source files (provided in step 5) to get an idea of how the development experience would look like.

1. Make sure you understand the whole setup by reading the [confluence document](https://liferay.atlassian.net/wiki/spaces/ENGFRONTENDINFRA/pages/2218164722/LPS-169939+-+Research+ways+to+use+Client+extensions+as+dependencies+of+other+extensions) describing it.
2. Ant all this PR.
3. Follow the steps in the [last commit](https://github.com/liferay-frontend/liferay-portal/pull/2967/commits/4df85cbaf48ff0e0ff1eb005799e4d278cd12d45) to install a modified JAR of `gradle-plugins-workspace` to your `liferay-portal/.m2` maven repo.
4. Create a local blade workspace and point its `settings.gradle` to your local `liferay-portal/.m2` maven repo as explained in the [last commit](https://github.com/liferay-frontend/liferay-portal/pull/2967/commits/4df85cbaf48ff0e0ff1eb005799e4d278cd12d45) to make it use the modifier JAR file created in step 3.
5. Extract the provided [greeter-and-provider.tar.gz](https://github.com/liferay-frontend/liferay-portal/files/10435106/greeter-and-provider.tar.gz) file in the `modules` folder of your blade workspace so that it creates projects `greeter` and `provider`.
6. Deploy those two projects to your local Liferay instance.
7. Add the `greeter` portlet to the home page and see it work (provided your browser supports import maps).
8. Bonus: use an older version (107.0.1 or less) of Firefox that doesn't support import maps to see how the shim makes it work even in that case. For this you need to:
    1. [Download Firefox 107](https://support.mozilla.org/en-US/kb/install-older-version-firefox).
    2. Open the portal (home page won't show the `greeter` portlet and JS console will say the browser cannot resolve `provider`).
    3. Go to `System Settings > Infrastructure > JavaScript Import Maps` and turn on `Enable ES Module Shims`.
    4. Refresh the home page (this time it should work).
